### DISCUSSION TITLE
Length of plot title string changes column widths in a layout

### DISCUSSION THREAD

**Discussion by ccarmen1** [pyqtgraph/pyqtgraph#2244] [Q&A]
I am trying create a GUI that contains 4 equally-sized plots in a 2x2 grid.  If the plots do not have titles, it works fine.  Once I add titles (if the title strings have different lengths), the column widths of the layout grid are no longer equal.  The code below is a minimal example that shows this behavior:

```python
import numpy as np
import pyqtgraph as pg

app = pg.mkQApp()
view = pg.GraphicsView()
l = pg.GraphicsLayout(border=(100, 100, 100))
view.setCentralItem(l)
view.resize(800, 600)

# Title strings are different lengths --> columns in the layout
# are different widths
p1 = l.addPlot(title="First Plot")
p2 = l.addPlot(title="Second Plot")

view.show()


if __name__ == '__main__':
    pg.exec()
```

The above code produces the following window on my computer, where the two plots are not the same width:
![Screen Shot 2022-03-31 at 11 13 46 AM](https://user-images.githubusercontent.com/102745419/161089596-ac979a91-b0c7-43d7-94e7-10776251bd25.png)

I have tried using a `GraphicsLayoutWidget` instead, and got the same results.  I also tried adding the title strings as independent labels in a row above the plots, and found the same issue.  

I am working on an Intel-based Macbook Pro, using python 3.8.  I setup a virtual environment and added only `PySide6` and `pyqtgraph`:
```
numpy==1.22.3
pyqtgraph==0.12.4
PySide6==6.2.4
shiboken6==6.2.4
```

Anyone have any ideas as to why this is happening, or what I can do to ensure the plots remain equal in width?

**Comment by j9ac9k** [comment]
Hi @ccarmen1 

We have an open issue to this, https://github.com/pyqtgraph/pyqtgraph/issues/1187  perhaps we should take another look here again.

I'm going to edit your post to get the syntax highlighting for your example code (thanks for providing!)

**Comment by ccarmen1** [comment]
Thanks @j9ac9k,

I did search the issues, but didn't see the item you referenced.  Too bad there isn't a solution, yet...

Also, thanks for updating my post to add syntax highlighting.  I'm new to github, and didn't know about that feature!

**Reply by j9ac9k** [reply]
> Also, thanks for updating my post to add syntax highlighting. I'm new to github, and didn't know about that feature!

github rolls out new features regularly, even if you're not new to github it's tough to keep up.  

We're about to release 0.12.5, which is a relatively small release but fixes a regression we introduced in 0.12.3 or 0.12.4.  If I get time I'll try and investigate this a  bit more closely before the release and see if there is a straight forward fix here.

**Comment by ccarmen1** [comment]
Given that there is still an open issue on this from 2020, I thought I'd look into a workaround that does not use the `GraphicsLayout` or `GraphicsLayoutWidget` to manage the grid of plots.  

I was able to achieve the desired result using a PySide6 `QGridLayout` object as the main layout.  For each plot I want to add, I create a `GraphicsView` instance, set its layout to a new `GraphicsLayout` instance, and put the `PlotItem` into that layout as its only child.  This way, the `QGridLayout` instance is controlling the width of the columns, and each `GraphicsLayout` instance is ensuring that the `PlotItem` takes up the entire region allocated to it.

The new version requires a bit more boilerplate to add all of the components for each new plot, but it does work.  Here's my new sample code to demonstrate this:

```python
import sys
from pyqtgraph import GraphicsView, GraphicsLayout
from PySide6.QtWidgets import QMainWindow, QApplication, QGridLayout, QWidget


class SingleDisplay(GraphicsView):
    def __init__(self, title_str):
        super().__init__()
        self.layout = GraphicsLayout()
        self.plot = self.layout.addPlot()
        self.plot.setTitle(title_str)
        self.plot.setLabel('left', "Y Axis")
        self.plot.setLabel('bottom', "X Axis")
        self.setCentralItem(self.layout)


class Window(QMainWindow):
    def __init__(self):
        super().__init__()

        grid = QGridLayout()
        grid.addWidget(SingleDisplay('First Plot'), 0, 0)
        grid.addWidget(SingleDisplay('Second Plot'), 0, 1)

        widget = QWidget()
        widget.setLayout(grid)
        widget.setStyleSheet("background-color:black")

        self.setGeometry(100, 100, 1000, 400)
        self.setCentralWidget(widget)
        self.show()


app = QApplication(sys.argv)
w = Window()
app.exec()
```

And the resulting window (where columns have equal length) looks like:
![Screen Shot 2022-04-01 at 1 32 42 PM](https://user-images.githubusercontent.com/102745419/161313504-10f5e834-ad38-4675-a248-08444ef7768c.png)

**Reply by j9ac9k** [reply]
@ccarmen1 I appreciate you posting a working solution here.  Thank you!

`pg.GraphicsLayout` inherits from `pg.GraphicsWidget`, and uses a `QGraphicsGridLayout` internally, as long as the issue isn't with the difference between `QGraphicsGridLayout` and `QGridLayout`, there is likely a fix here that won't be too difficult to spot.

Regarding the duration for which the issue was open; yeah it's a problem.  I have directed my effort to address regressions and minimize the outstanding PR count.  The issue tracker on its own doesn't get as much attention as it should.

We could certainly use more maintainer hours to address existing open issues.  I actually posted about this on the mail list a little over a year ago [here](https://groups.google.com/g/pyqtgraph/c/uifEdGPFzXI/m/ZT76F5wvBAAJ).  If you (or anyone else reading this) have an interest in contributing to the library on a somewhat regular basis, please feel free to reach out.  My email is posted on my github profile; or you can tweet at `@pyqtgraph`.  The barrier to entry is not particularly high.  I discuss this a bit more in the linked mail list post.

**Comment by pijyoi** [comment]
This is perhaps also related to #2107. (closed but not really resolved)

**Comment by dziyaee** [comment]
It's curious that the `pg.PlotItem` with the shorter title is given more space. Not sure what's causing this, but I solved this by setting the sizeHint of one `pg.PlotItem` equal to the other. 

```python
import pyqtgraph as pg
from PyQt5.QtCore import QSizeF


class LinkedPlot(pg.PlotItem):
    def __init__(self, *args, otherPlot, **kwargs):
        super().__init__(*args, **kwargs)
        self.otherPlot = otherPlot

    def sizeHint(self, which, constraint = QSizeF()):
        try:
            return self.otherPlot.sizeHint(which, constraint)

        except AttributeError:
            return super().sizeHint(which, constraint)


app = pg.mkQApp()
view = pg.GraphicsView()
l = pg.GraphicsLayout(border=(100, 100, 100))
view.setCentralItem(l)
view.resize(800, 600)

# Title strings are different lengths --> columns in the layout
# are different widths
p1 = pg.PlotItem(title = "First Plot")
p2 = LinkedPlot(title = "Second Plot", otherPlot = p1)
l.addItem(p1)
l.addItem(p2)

view.show()

if __name__ == '__main__':
    app.exec()

```

Result:
<img width="912" alt="Screen Shot 2022-04-05 at 8 11 56 AM" src="https://user-images.githubusercontent.com/13340419/161786569-081ad07c-377e-46d3-b7db-a8794906cb6f.png">

**Reply by ccarmen1** [reply]
@dziyaee - Thanks for looking into this issue some more.  It's interesting that overloading the size hints in the PlotItems can resolve the problem.  It seems to me that there is some interplay between the LabelItem that holds the title string and the size hint for the PlotItem.  I found that I could also address this setting the size policy for each PlotItem to the QSizePolicy.Ignored state, instead of overloading the size hints:

```python
import pyqtgraph as pg
from PySide6.QtCore import QSizeF
from PySide6.QtWidgets import QSizePolicy

app = pg.mkQApp()
view = pg.GraphicsView()
l = pg.GraphicsLayout(border=(100, 100, 100))
view.setCentralItem(l)
view.resize(800, 600)

p1 = pg.PlotItem(title="First Plot")
p1.setSizePolicy(QSizePolicy.Ignored, QSizePolicy.Preferred)
p2 = pg.PlotItem(title="Second Plot")
p2.setSizePolicy(QSizePolicy.Ignored, QSizePolicy.Preferred)

l.addItem(p1)
l.addItem(p2)

view.show()

if __name__ == '__main__':
    app.exec()
```

Like the other approaches, this works, but is somewhat unsatisfying.  It does nothing to explain why the plot with the shorter title string is given more space, not does it explain why the title string's sizes are driving the width of the PlotItem, when the title string is clearly much smaller than the available space...
